### ISSUE TITLE
[Usage]: Which branch should I use to test speculative decoding

### ISSUE THREAD

**Issue by v-lmn** [vllm-project/vllm#9400]
### Your current environment

```text
The output of `python collect_env.py`
```

### How would you like to use vllm

Which branch should I use to test speculative decoding, and which branch currently has the best optimizations for speculative decoding? Is it the main branch?

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by LiuXiaoxuanPKU** [comment]
Yes please use the main branch. The draft model based method and ngram should have reasonable performance. Medusa/Eagle might have worse performance than expected.

**Comment by v-lmn** [comment]
> Yes please use the main branch. The draft model based method and ngram should have reasonable performance. Medusa/Eagle might have worse performance than expected.

佬，你看看我用ngram的启动命令有没有问题，为什么比draft model based的方式也没快很多呀
CUDA_VISIBLE_DEVICES=7 python -m vllm.entrypoints.openai.api_server --model /mnt/nvme0n1/ckpt/llama/Meta-Llama-3.1-8B-Instruct --trust-remote-code -tp 1 --speculative_model [ngram] --num_speculative_tokens 5 --ngram_prompt_lookup_max 4 --seed 42 --use-v2-block-manager --chat-template ./examples/template_chatml.jinja --host 0.0.0.0 --port 9001

**Comment by v-lmn** [comment]
> > Yes please use the main branch. The draft model based method and ngram should have reasonable performance. Medusa/Eagle might have worse performance than expected.
> 
> 佬，你看看我用ngram的启动命令有没有问题，为什么比draft model based的方式也没快很多呀 CUDA_VISIBLE_DEVICES=7 python -m vllm.entrypoints.openai.api_server --model /mnt/nvme0n1/ckpt/llama/Meta-Llama-3.1-8B-Instruct --trust-remote-code -tp 1 --speculative_model [ngram] --num_speculative_tokens 5 --ngram_prompt_lookup_max 4 --seed 42 --use-v2-block-manager --chat-template ./examples/template_chatml.jinja --host 0.0.0.0 --port 9001 @LiuXiaoxuanPKU

CUDA_VISIBLE_DEVICES=0 python -m vllm.entrypoints.openai.api_server --model /mnt/nvme0n1/ckpt/llama/Meta-Llama-3.1-8B-Instruct --trust-remote-code -tp 1 --speculative_model [ngram] --num_speculative_tokens 4 --ngram-prompt-lookup-max 10 --seed 42 --use-v2-block-manager --host 0.0.0.0 --port 9001
在h100单卡上性能提升约35%

**Comment by LiuXiaoxuanPKU** [comment]
It really depends on the workload you are using. Ngram only works when there are big overlap between prompts and answers, such as summarization workloads. It generally does not work well for chatting dataset.

**Comment by bettybaii** [comment]
> It really depends on the workload you are using. Ngram only works when there are big overlap between prompts and answers, such as summarization workloads. It generally does not work well for chatting dataset.

Thank you so much, [@LiuXiaoxuanPKU](https://github.com/LiuXiaoxuanPKU), for sharing your insights! Could you kindly recommend any datasets where Ngram performs well and meets the “big overlap between prompts and answers” characteristic you mentioned? Any suggestions would be greatly appreciated.

**Comment by LiuXiaoxuanPKU** [comment]
> > It really depends on the workload you are using. Ngram only works when there are big overlap between prompts and answers, such as summarization workloads. It generally does not work well for chatting dataset.
> 
> Thank you so much, [@LiuXiaoxuanPKU](https://github.com/LiuXiaoxuanPKU), for sharing your insights! Could you kindly recommend any datasets where Ngram performs well and meets the “big overlap between prompts and answers” characteristic you mentioned? Any suggestions would be greatly appreciated.

Try this: https://github.com/vllm-project/vllm/blob/8ec940fda8757cc2cfceb5913aaad7916528576b/benchmarks/benchmark_serving.py#L92

**Comment by bettybaii** [comment]
> > > It really depends on the workload you are using. Ngram only works when there are big overlap between prompts and answers, such as summarization workloads. It generally does not work well for chatting dataset.
> > 
> > 
> > Thank you so much, [@LiuXiaoxuanPKU](https://github.com/LiuXiaoxuanPKU), for sharing your insights! Could you kindly recommend any datasets where Ngram performs well and meets the “big overlap between prompts and answers” characteristic you mentioned? Any suggestions would be greatly appreciated.
> 
> Try this:
> 
> https://github.com/vllm-project/vllm/blob/8ec940fda8757cc2cfceb5913aaad7916528576b/benchmarks/benchmark_serving.py#L92

Thank you so much for your reply![@LiuXiaoxuanPKU](https://github.com/LiuXiaoxuanPKU) I truly appreciate it! 😊
